File: AI-Reply/Shared/Account/AccountService.swift
import Foundation

/// Every call the app makes to the AI Reply backend.
///
/// Барлық сұраныс осы жерден өтеді: қайталанатын желі коды жоқ.
///
/// The service holds no state of its own. Tokens belong to `AccountSession`,
/// and the base URL to `AIConfiguration`, so a screen only has to say what it
/// wants, not how authentication works.
struct AccountService: Sendable {

    private let session: AccountSession

    init(session: AccountSession = .shared) {
        self.session = session
    }

    private func client() throws -> APIClient {
        guard let baseURL = AIConfiguration.shared.backendBaseURL else {
            throw APIError.invalidRequest
        }
        return APIClient(baseURL: baseURL)
    }

    // MARK: Public endpoints

    /// Countries, limits and whether the demo OTP is on. Called before sign-in.
    func serverConfig() async throws -> AccountAPI.ServerConfig {
        try await client().get("api/v1/config")
    }

    func plans() async throws -> [AccountAPI.Plan] {
        let list: AccountAPI.PlanList = try await client().get("api/v1/plans")
        return list.plans
    }

    // MARK: Sign-in

    /// Asks for a code. The server decides how it is delivered - in demo mode
    /// nothing is sent at all and `demoMode` comes back true.
    func requestCode(identifier: String, locale: String) async throws -> AccountAPI.Challenge {
        struct Request: Encodable {
            let identifier: String
            let locale: String
        }
        return try await client().post("api/v1/auth/request-otp",
                                       body: Request(identifier: identifier, locale: locale))
    }

    /// Verifies the code and stores the resulting session.
    @discardableResult
    func verifyCode(identifier: String, code: String) async throws -> AccountAPI.Session {
        struct Request: Encodable {
            let identifier: String
            let code: String
            let device: DeviceDescriptor
        }
        let result: AccountAPI.Session = try await client().post(
            "api/v1/auth/verify-otp",
            body: Request(identifier: identifier, code: code, device: .current)
        )
        await session.adopt(result)
        return result
    }

    func signOut() async {
        await session.signOut()
    }

    // MARK: Authenticated endpoints

    func account() async throws -> AccountAPI.Account {
        try await session.authenticated { token in
            try await client().get("api/v1/me", token: token)
        }
    }

    func usage() async throws -> AccountAPI.Usage {
        try await session.authenticated { token in
            try await client().get("api/v1/me/usage", token: token)
        }
    }

    func subscription() async throws -> AccountAPI.Subscription {
        try await session.authenticated { token in
            try await client().get("api/v1/me/subscription", token: token)
        }
    }

    /// Sends only the fields that changed - every property is optional on the
    /// wire, so an empty edit is a no-op rather than an accidental wipe.
    struct ProfileUpdate: Encodable, Sendable {
        var display_name: String?
        var role: String?
        var description: String?
        var preferred_tone: String?
        var business_offering: String?
        var business_summary: String?
        var business_rules: [String]?
        var locale: String?
        var timezone: String?
        var onboarding_completed: Bool?
    }

    @discardableResult
    func updateProfile(_ update: ProfileUpdate) async throws -> AccountAPI.Profile {
        try await session.authenticated { token in
            try await client().patch("api/v1/me", body: update, token: token)
        }
    }

    /// Registers this device so a push token has somewhere to live later.
    func registerDevice(pushToken: String? = nil) async throws {
        struct Request: Encodable {
            let device_id: String
            let platform: String
            let app_version: String
            let os_version: String
            let locale: String
            let push_token: String?
            let push_enabled: Bool
        }
        let descriptor = DeviceDescriptor.current
        let request = Request(
            device_id: descriptor.device_id,
            platform: descriptor.platform,
            app_version: descriptor.app_version,
            os_version: descriptor.os_version,
            locale: descriptor.locale,
            push_token: pushToken,
            push_enabled: pushToken != nil
        )
        let _: APIClient.Empty = try await session.authenticated { token in
            try await client().post("api/v1/devices", body: request, token: token)
        }
    }

    // MARK: Subscription changes (demo payment adapter)

    struct CheckoutResult: Decodable, Sendable {
        let paymentID: String
        let provider: String
        let status: String
        let demo: Bool

        enum CodingKeys: String, CodingKey {
            case paymentID = "payment_id"
            case provider, status, demo
        }
    }

    func startCheckout(planID: String) async throws -> CheckoutResult {
        struct Request: Encodable { let plan_id: String }
        return try await session.authenticated { token in
            try await client().post("api/v1/payments/checkout",
                                    body: Request(plan_id: planID), token: token)
        }
    }

    /// Confirms a payment. With the demo adapter this is what actually moves
    /// the account onto the chosen plan; with a real acquirer the confirmation
    /// arrives from the provider instead and this call just reads the result.
    @discardableResult
    func confirmCheckout(paymentID: String) async throws -> AccountAPI.Usage {
        struct Response: Decodable, Sendable {
            let subscription: AccountAPI.Subscription
            let usage: AccountAPI.Usage
        }
        let response: Response = try await session.authenticated { token in
            try await client().post("api/v1/payments/\(paymentID)/confirm",
                                    body: APIClient.EmptyBody(), token: token)
        }
        return response.usage
    }
}

extension APIClient {
    /// A body for endpoints that take none but still expect JSON.
    struct EmptyBody: Encodable, Sendable {}
}

File: AI-Reply-Android/app/src/main/java/kz/yerek/aireply/data/account/AccountService.kt
package kz.yerek.aireply.data.account

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kz.yerek.aireply.data.legal.StoredLegalConsent

/**
 * What the client tells the server about itself.
 *
 * Құрылғы туралы ең қажетті мәлімет қана.
 *
 * Deliberately minimal: platform, app version, OS version, locale, time zone.
 * No device name, no model identifier, no ANDROID_ID, no advertising id.
 */
@Serializable
data class DeviceDescriptor(
    @SerialName("device_id") val deviceId: String,
    val platform: String,
    @SerialName("app_version") val appVersion: String,
    @SerialName("os_version") val osVersion: String,
    val locale: String,
    val timezone: String
)

@Serializable
internal data class RefreshRequest(
    @SerialName("refresh_token") val refreshToken: String,
    val device: DeviceDescriptor
)

@Serializable
internal data class LogoutRequest(@SerialName("refresh_token") val refreshToken: String)

@Serializable
private data class RequestOtpRequest(val identifier: String, val locale: String)

@Serializable
private data class VerifyOtpRequest(
    val identifier: String,
    val code: String,
    val device: DeviceDescriptor
)

/** Only the fields that changed; everything else is left alone server-side. */
@Serializable
data class ProfileUpdate(
    @SerialName("display_name") val displayName: String? = null,
    val role: String? = null,
    val description: String? = null,
    @SerialName("preferred_tone") val preferredTone: String? = null,
    @SerialName("business_offering") val businessOffering: String? = null,
    @SerialName("business_summary") val businessSummary: String? = null,
    @SerialName("business_rules") val businessRules: List<String>? = null,
    val locale: String? = null,
    val timezone: String? = null,
    @SerialName("onboarding_completed") val onboardingCompleted: Boolean? = null
)

@Serializable
private data class RegisterDeviceRequest(
    @SerialName("device_id") val deviceId: String,
    val platform: String,
    @SerialName("app_version") val appVersion: String,
    @SerialName("os_version") val osVersion: String,
    val locale: String,
    @SerialName("push_token") val pushToken: String? = null,
    @SerialName("push_enabled") val pushEnabled: Boolean
)

@Serializable
private data class CheckoutRequest(@SerialName("plan_id") val planId: String)

@Serializable
private data class LegalConsentRequest(
    @SerialName("terms_version") val termsVersion: String,
    @SerialName("privacy_version") val privacyVersion: String,
    val locale: String,
    val platform: String,
    @SerialName("app_version") val appVersion: String
)

/**
 * Every call the app makes to the AI Reply backend.
 *
 * Барлық сұраныс осы жерден өтеді: қайталанатын желі коды жоқ.
 *
 * The service holds no state. Tokens belong to [AccountSession] and the base
 * URL to the configuration, so a screen only says what it wants, not how
 * authentication works.
 */
class AccountService(
    private val session: AccountSession,
    private val baseUrlProvider: () -> String?,
    private val deviceDescriptor: () -> DeviceDescriptor
) {

    private fun client(): ApiClient {
        val baseUrl = baseUrlProvider() ?: ApiError.InvalidRequest.raise()
        return ApiClient(baseUrl)
    }

    private val json: Json get() = jsonCodec

    // ------------------------------------------------------ public endpoints

    /** Countries, limits and current legal versions. Called before sign-in. */
    suspend fun serverConfig(): ServerConfigDto {
        val client = client()
        return decode(ServerConfigDto.serializer(), client.request("GET", "api/v1/config"))
    }

    suspend fun plans(): List<PlanDto> {
        val client = client()
        return decode(PlanListDto.serializer(), client.request("GET", "api/v1/plans")).plans
    }

    // -------------------------------------------------------------- sign-in

    /** Asks for a code. The server decides how it is delivered. */
    suspend fun requestCode(identifier: String, locale: String): ChallengeDto {
        val client = client()
        val body = json.encodeToString(
            RequestOtpRequest.serializer(), RequestOtpRequest(identifier, locale)
        )
        return decode(ChallengeDto.serializer(), client.request("POST", "api/v1/auth/request-otp", body))
    }

    /** Verifies the code and stores the resulting session. */
    suspend fun verifyCode(identifier: String, code: String): AccountSessionDto {
        val client = client()
        val body = json.encodeToString(
            VerifyOtpRequest.serializer(),
            VerifyOtpRequest(identifier, code, deviceDescriptor())
        )
        val result = decode(
            AccountSessionDto.serializer(),
            client.request("POST", "api/v1/auth/verify-otp", body)
        )
        session.adopt(result)
        return result
    }

    suspend fun signOut() = session.signOut()

    // ----------------------------------------------- authenticated endpoints

    suspend fun account(): AccountDto = session.authenticated { token ->
        decode(AccountDto.serializer(), client().request("GET", "api/v1/me", token = token))
    }

    suspend fun usage(): UsageDto = session.authenticated { token ->
        decode(UsageDto.serializer(), client().request("GET", "api/v1/me/usage", token = token))
    }

    suspend fun subscription(): SubscriptionDto = session.authenticated { token ->
        decode(SubscriptionDto.serializer(), client().request("GET", "api/v1/me/subscription", token = token))
    }

    /**
     * Saves profile changes.
     *
     * POST rather than PATCH on purpose: HttpURLConnection refuses PATCH, and
     * the server exposes this alias for exactly that reason.
     */
    suspend fun updateProfile(update: ProfileUpdate): AccountProfile = session.authenticated { token ->
        val body = json.encodeToString(ProfileUpdate.serializer(), update)
        decode(AccountProfile.serializer(), client().request("POST", "api/v1/me", body, token))
    }

    /** Registers this device so a push token has somewhere to live later. */
    suspend fun registerDevice(pushToken: String? = null) {
        val descriptor = deviceDescriptor()
        val body = json.encodeToString(
            RegisterDeviceRequest.serializer(),
            RegisterDeviceRequest(
                deviceId = descriptor.deviceId,
                platform = descriptor.platform,
                appVersion = descriptor.appVersion,
                osVersion = descriptor.osVersion,
                locale = descriptor.locale,
                pushToken = pushToken,
                pushEnabled = pushToken != null
            )
        )
        session.authenticated { token ->
            client().request("POST", "api/v1/devices", body, token)
        }
    }

    suspend fun recordLegalConsent(consent: StoredLegalConsent): LegalConsentDto =
        session.authenticated { token ->
            val request = LegalConsentRequest(
                termsVersion = consent.termsVersion,
                privacyVersion = consent.privacyVersion,
                locale = consent.locale,
                platform = consent.platform,
                appVersion = consent.appVersion
            )
            val body = json.encodeToString(LegalConsentRequest.serializer(), request)
            decode(
                LegalConsentDto.serializer(),
                client().request("POST", "api/v1/me/consents", body, token)
            )
        }

    // --------------------------------------- subscription (demo payment flow)

    suspend fun startCheckout(planId: String): CheckoutDto = session.authenticated { token ->
        val body = json.encodeToString(CheckoutRequest.serializer(), CheckoutRequest(planId))
        decode(CheckoutDto.serializer(), client().request("POST", "api/v1/payments/checkout", body, token))
    }

    /**
     * Confirms a payment. With the demo adapter this is what actually moves the
     * account onto the chosen plan; with a real acquirer the confirmation
     * arrives from the provider and this call only reads the result.
     */
    suspend fun confirmCheckout(paymentId: String): CheckoutResultDto = session.authenticated { token ->
        decode(
            CheckoutResultDto.serializer(),
            client().request("POST", "api/v1/payments/$paymentId/confirm", "{}", token)
        )
    }

    private companion object {
        val jsonCodec = Json {
            ignoreUnknownKeys = true
            encodeDefaults = false
            explicitNulls = false
        }

        /** A body we cannot parse is a malformed response, not a crash. */
        fun <T> decode(serializer: kotlinx.serialization.KSerializer<T>, body: String): T =
            runCatching { jsonCodec.decodeFromString(serializer, body) }
                .getOrElse { ApiError.MalformedResponse.raise() }
    }
}
